man/man8/puppetca.8
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "PUPPETCA" "8" "August 2010" "" ""
Stand\-alone certificate authority\. Capable of generating certificates but mostly meant for signing certificate requests from puppet clients\.puppet cert [\-h|\-\-help] [\-V|\-\-version] [\-d|\-\-debug] [\-v|\-\-verbose]
.
.IP "" 4
.
.nf

          [\-g|\-\-generate] [\-l|\-\-list] [\-s|\-\-sign] [\-r|\-\-revoke]
          [\-p|\-\-print] [\-c|\-\-clean] [\-\-verify] [\-\-digest DIGEST]
          [\-\-fingerprint] [host]
.
.fi
.
.IP "" 0
Because the puppetmasterd daemon defaults to not signing client certificate requests, this script is available for signing outstanding requests\. It can be used to list outstanding requests and then either sign them individually or sign all of them\.Note that any configuration parameter that\'s valid in the configuration file is also a valid long argument\. For example, \'ssldir\' is a valid configuration parameter, so you can specify \'\-\-ssldir \fIdirectory\fR\' as an argument\.
.
.P
See the configuration file documentation at http://reductivelabs\.com/projects/puppet/reference/configref\.html for the full list of acceptable parameters\. A commented list of all configuration options can also be generated by running puppet cert with \'\-\-genconfig\'\.
.
.P
all: Operate on all items\. Currently only makes sense with
.
.IP "" 4
.
.nf

         \'\-\-sign\', \'\-\-clean\', or \'\-\-list\'\.
.
.fi
.
.IP "" 0
.
.P
digest: Set the digest for fingerprinting (defaults to md5)\. Valid
.
.IP "" 4
.
.nf

         values depends on your openssl and openssl ruby extension
         version, but should contain at least md5, sha1, md2,
         sha256\.
.
.fi
.
.IP "" 0
.
.P
clean: Remove all files related to a host from puppet cert\'s
.
.IP "" 4
.
.nf

         storage\. This is useful when rebuilding hosts, since new
         certificate signing requests will only be honored if puppet
         cert does not have a copy of a signed certificate for that
         host\. The certificate of the host remains valid\. If \'\-\-all\'
         is specified then all host certificates, both signed and
         unsigned, will be removed\.
.
.fi
.
.IP "" 0
.
.P
debug: Enable full debugging\.
.
.P
generate: Generate a certificate for a named client\. A
.
.IP "" 4
.
.nf

         certificate/keypair will be generated for each client named
         on the command line\.
.
.fi
.
.IP "" 0
.
.P
help: Print this help message
.
.P
list: List outstanding certificate requests\. If \'\-\-all\' is
.
.IP "" 4
.
.nf

         specified, signed certificates are also listed, prefixed by
         \'+\', and revoked or invalid certificates are prefixed by
         \'\-\' (the verification outcome is printed in parenthesis)\.
.
.fi
.
.IP "" 0
.
.P
print: Print the full\-text version of a host\'s certificate\.
.
.P
fingerprint: Print the DIGEST (defaults to md5) fingerprint of a host\'s
.
.IP "" 4
.
.nf

         certificate\.
.
.fi
.
.IP "" 0
.
.P
revoke: Revoke the certificate of a client\. The certificate can be
.
.IP "" 4
.
.nf

         specified either by its serial number, given as a decimal
         number or a hexadecimal number prefixed by \'0x\', or by its
         hostname\. The certificate is revoked by adding it to the
         Certificate Revocation List given by the \'cacrl\' config
         parameter\. Note that the puppetmasterd needs to be
         restarted after revoking certificates\.
.
.fi
.
.IP "" 0
.
.P
sign: Sign an outstanding certificate request\. Unless \'\-\-all\' is
.
.IP "" 4
.
.nf

         specified, hosts must be listed after all flags\.
.
.fi
.
.IP "" 0
.
.P
verbose: Enable verbosity\.
.
.P
version: Print the puppet version number and exit\.
.
.P
verify: Verify the named certificate against the local CA
.
.IP "" 4
.
.nf

         certificate\.
.
.fi
.
.IP "" 0
$ puppet cert \-l culain\.madstop\.com $ puppet cert \-s culain\.madstop\.comLuke KaniesCopyright (c) 2005 Puppet Labs, LLC Licensed under the GNU Public License


CHANGELOG
2.6.3
=====
184733c  [#5322] (#5322) Remove spec file that adds little value and causes failures
178c2a6  Fix test failures that fixing #4726 exposed.
74b6c09  (#4726) Fix RRD legacy support
8662056  Fix for #4279 -- mount detection on HP-UX
fbb096a  Fix for #5055 -- adding to_sym to Puppet::Node::Environment
b2ff6a5  Fix for #5298 -- Collections need to do type lookup
1ce00dc  Step towards [5298] -- cleanup indentation, etc. in AST::Collection
722608b  Fixed #5287 - Schedule documentation is incorrect
c8b6fb5  Fixed #5296 - test warnings messages
d221c05  (#5297) Fix schedule tests that were missing stubs for Time.now
f2fd0d1  Fix for #5273 -- user resource constantly resetting password age
544dcf8  Fix #5289 -- Bad copy/paste changes message on test failure

2.6.3rc3
========
126681f  Updated CHANGELOG for 2.6.3rc3
b15231d  Fix for #4299 -- Don't require which
ea435a4  Fix #5020 - Prefer finding node name from REST uri over certname
a097b93  Fix for #4894 -- retry tests if port is in use
ee61b4e  Fix for #4955 -- Race condition & memory leak in Puppet::Util
f57425d  Fix #4921 - race condition in Parser Functions creation
9604f1c  Fix #5252 - line number mis-attribution during parsing
cc5224c  Maint. fix for test broken by 00eedac5
5f7d0fb  Fix for #2568 -- Add a dbconnections option to set AR pool size
ba4d22b  Maint.  Removing code for which no CLA has been signed
4a3d5d7  Reimplementation of functionality removed by prior commit
235d641  Refactor for CLA
9ba0c8a  Fix #4923 - close process race when truncating existing file
cb16d3d  Puppet-load: better and safer error reporting
1d26742  Fix #5023 - puppet-load multiple nodes support
00eedac  capture stderr from exec resources
4cbceab  (#4573) FreeBSD service provider now supports versions <7 and >8
06c8748  Fix #3808 - puppetdoc should use --force-update only if RDoc supports it
6e6712b  [#4813] Remove dead code from puppet/rails/host.rb
956296a  Fix #4911 - Do not generate doc for standard RDoc parser generated object
4fa24bb  Fix #5127 - error when accessing array elements
abb8c66  (#5242) Fix schedule specs that fail near daylight savings
ec667fd  Kludge for #5206 -- port of fix for #3536 to yaml
9a3b584  (#5062) Add envpuppet helper script to ext/
aad7008  [#5225] Fix spec failure that depended on time change
21db472  (#5233) Randomize tmp dir paths

2.6.3rc2
========
244213c  Updated CHANGELOG for 2.6.3rc2
76ac1f8  Fixed #5112 - Launchd Service broke in 2.6.2 with OS X 10.4 Clients.
776ea2a  Fixed #5137 - Removed no longer required TOC references
31118fe  Kludge for #5048 -- serialization compatibility with 0.25.x
65ef24e  (#4534/#4778) -- Normalize parameterized classes
3b53bfc  Fix for #5022 -- Escaped newlines should be elided

2.6.3rc1
========
e3fc5b9  Updated CHANGELOG and version for 2.6.3rc1
3c56705  Fix for #4832 -- Making PSON handle arbitrary binary data
e232770  Minimal fix for #4975 -- only call chage when managing password age rules
a090e86  Fix for #4963 -- Use correct commands for password expiry on solaris

2.6.2
=====
295c3be  Updated CHANGELOG for 2.6.2
1d719be  Fix for #4945 -- explicitly check os to supress man page installation
55417bc  Reversion of 021d534482dd8edb863cb77d668ac3525362a0a6
a6e2bea  Fixed #4919 - added parenths to fix error message:

2.6.2rc1
========
917c520  Incremented version to 2.6.2
900eae4  Updated CHANGELOG for 2.6.2rc1
1b6094d  Fixed documentation typo
bdf12fe  Fix for #4896 -- stray newline left over from removed diagnostic
e7424c6  (#4772) Update SuSE .spec file
0aaa742  Fixes #4792 (Duplicate definition since 2.6.1 upgrade)
ea49d13  Improvement to #4025: made spec tests work on all platforms
0b4ce08  Adds #3046 - support for password min/max age
e9f9d26  [#4783] (#4783) Fix RRDGraph report generation
34f87cf  Add user account expiry to the useradd type and provider
a7fb9b1  Fixed #4025 (failure in launchd if certain plists are binary).
2573872  Fix for #4649 -- avoid unsupported features on non-posix systems
eb9279c  Fix for 4273 -- revert b7e2580ab49ecdb67fc9b522829c005fc3750fbe
53a2bea  Fix for #4804 -- escaped backslashes in interpolated strings
d12e477  Fixes #4863 (Missing "require 'webrick'" causes nondeterministic spec failures)
574812e  (#4860) Add regression tests that would have caught bad params method
68947e7  (#4860) Fix wrong method name.. params seems to be renamed to parameters
021359b  Fix for #4644: install.rb works properly on Windows
d057b90  Fix #4726 Update puppet rrdtool metric code to support modern rrd ruby bindings
66cf3a9  Fix #4226 - Prepend 'Puppet CA: ' to fqdn for default root ca_name
d54352a  Port Puppet::SSLCertificates::CA test to rspec
effc6b8  Fixes #4852 - error messages involving Whits now mention Classes instead
3f99bd7  Fix #4267 - Create a backup before dropping permissions
6468f4e  (#4763) Don't call a method that was removed in Rails 3 activerecord
79d5fde  Fixed #4763 - Hardcoded ActiveRecord version
4798df3  Fixes #4822 -- Puppet doc -o option broken
99c1019  [#4798] Puppet doc manifests documentation mode broken
8cd1540  [#4692] undefined variables cause :undef to be passed to functions
06bf566  [#4787] Missing require causing failure
bba04e0  Fix for #4746 -- Newline goes at the _end_ of the line
9e17c25  Fix #4743: Allow the audit meta-parameter to accept both 'all', and :all
f950061  [#4716] ResourceTypeAPI exposes implementation details that are likely to change
8ff4b9a  Fixed #4819 - corrected cron documentation
2b50f30  [#4771] Import of manifests with the same name only happens once
7b8cb74  Fix for #4708 - tagmail should allow . in tagname
6f229ee  Minimal fix for #4631 -- set implicit classes as in 0.25.x
021d534  Fixed #3707 - rpm, like dpkg-query exits 1 if the package is not installed.  Returning nil in this provider had the effect that on every run, puppet would end up calling yum erase .  Returning the correct data structure resolves this.
216f663  Fixed Puppet Doc TOC generation
c3cb57c  Fixed versioncmp function typo

lib/puppet/application/puppetca.rb
require 'puppet'
require 'puppet/application'
require 'puppet/ssl/certificate_authority'

Puppet::Application.new(:puppetca) do

    should_parse_config

    attr_accessor :cert_mode, :all, :signed, :ca

    def find_mode(opt)
        modes = Puppet::SSL::CertificateAuthority::Interface::INTERFACE_METHODS
        tmp = opt.sub("--", '').to_sym
        @cert_mode = modes.include?(tmp) ? tmp : nil
    end

    option("--clean", "-c") do
        @cert_mode = :destroy
    end

    option("--all", "-a") do
        @all = true
    end

    option("--signed", "-s") do
        @signed = true
    end

    option("--debug", "-d") do |arg|
        Puppet::Util::Log.level = :debug
    end

    Puppet::SSL::CertificateAuthority::Interface::INTERFACE_METHODS.reject {|m| m == :destroy }.each do |method|
        option("--#{method}", "-%s" % method.to_s[0,1] ) do
            find_mode("--#{method}")
        end
    end

    option("--verbose", "-v") do
        Puppet::Util::Log.level = :info
    end

    command(:main) do
        if @all
            hosts = :all
        elsif @signed
            hosts = :signed
        else
            hosts = ARGV.collect { |h| puts h; h.downcase }
        end
        begin
            @ca.apply(@cert_mode, :to => hosts)
        rescue => detail
            puts detail.backtrace if Puppet[:trace]
            puts detail.to_s
            exit(24)
        end
    end

    setup do
        if Puppet.settings.print_configs?
            exit(Puppet.settings.print_configs ? 0 : 1)
        end

        Puppet::Util::Log.newdestination :console

        if [:generate, :destroy].include? @cert_mode
            Puppet::SSL::Host.ca_location = :local
        else
            Puppet::SSL::Host.ca_location = :only
        end

        begin
            @ca = Puppet::SSL::CertificateAuthority.new
        rescue => detail
            puts detail.backtrace if Puppet[:trace]
            puts detail.to_s
            exit(23)
        end
    end
end
